File: Iterators/IteratorOfCombinationCpp/IteratorOfCombination.cpp
#include "IteratorOfCombination.h"

using namespace std;

IteradorCombinaciones::IteradorCombinaciones(int N) {
    valorMaximo = 1;
    valorMinimo = 0;
    tamano = N;
    combinacion = new int[tamano];
    for (int i = 0; i < N; i++) {
        combinacion[i] = 0;
    }
}

bool IteradorCombinaciones::ultimaCombinacion() {
    for (int i = 0; i < tamano; i++) {
        if (combinacion [i] == 0) {
            return false;
        }
    }
    return true;
}

int* IteradorCombinaciones::siguienteCombinacion() {
    for (int i = 0; i < tamano; i++) {
        if (combinacion[i] == valorMaximo) {
            combinacion[i] = valorMinimo;
        } else {
            combinacion[i] = combinacion[i] + 1;
            return combinacion;
        }
    }
    return combinacion;
}

File: Iterators/IteratorOfCombinationJava/src/iteratorofcombination/IteratorOfCombination.java
package iteratorofcombination;

class IteratorOfCombination {

    private int[] combinacion;
    private int valormaximo;
    private int valorminimo;

    public IteratorOfCombination(int N) {
        this.combinacion = new int[N];
        this.valormaximo = 1;
        this.valorminimo = 0;
    }

    public boolean ultimaCombinacion() {
        for (int i = 0; i < combinacion.length; i++) {
            if (combinacion[i] == 0) {
                return false;
            }
        }
        return true;
    }

    public int[] siguienteCombinacion() {
        for (int i = 0; i < combinacion.length; i++) {
            if (combinacion[i] == valormaximo) {
                combinacion[i] = valorminimo;
            } else {
                combinacion[i] = combinacion[i] + 1;
                return combinacion;
            }
        }
        return null;
    }
}
